proxy/envoy/ingress.go

```diff
@@ -0,0 +1,244 @@
+	}
+
+	// TODO: add retry logic
+	if err := w.agent.Reload(config); err != nil {
+		glog.Warningf("Envoy reload error: %v", err)
+		return
+	}
+
+	// Add a short delay to de-risk a potential race condition in envoy hot reload code.
+	// The condition occurs when the active Envoy instance terminates in the middle of
+	// the Reload() function.
+	time.Sleep(256 * time.Millisecond)
+}
+
+func (w *ingressWatcher) generateConfig() (*Config, error) {
+	// Map a domain name to a VirtualHost
+	vhostsMap := make(map[string]VirtualHost)
+
+	// Map a cluster name to a Cluster
+	clustersMap := make(map[string]Cluster)
+
+	// Function to add a rule for a particular host, path and backend.
+	// This will later be invoked for each ingress rule, as well as ingress default backend
+	addRule := func(host, path, namespace string, backend *v1beta1.IngressBackend) {
+
+		// If host is unspecified, use Envoy's wildcard domain "*"
+		domain := host
+		if domain == "" {
+			domain = "*"
+		}
+
+		// Fetch domain's VirtualHost, or create one if doesn't exist yet
+		vhost, ok := vhostsMap[domain]
+		if !ok {
+			vhost := VirtualHost{
+				Name:    domain,
+				Domains: []string{domain},
+			}
+			vhostsMap[domain] = vhost
+		}
+
+		// Fetch target cluster, or create one if doesn't exist yet
+		clusterName := clusterName(namespace, backend)
+		cluster, ok := clustersMap[clusterName]
+		if !ok {
+			clustersMap[clusterName] = Cluster{
+				Name:             clusterName,
+				ServiceName:      clusterName,
+				Type:             "strict_dns", // TODO: Change to SDS
+				LbType:           DefaultLbType,
+				ConnectTimeoutMs: DefaultTimeoutMs,
+				Hosts:            clusterHosts(namespace, backend),
+			}
+		}
+
+		route := Route{
+			Cluster: cluster.Name,
+		}
+		vhost.Routes = append(vhost.Routes, route)
+
+		// TODO: Support path regexes
+		if path == "" {
+			route.Prefix = "/"
+		} else {
+			route.Path = path
+		}
+
+		// Ensure "catch-all" path appears last on the ordered slice of routes
+		if rLen := len(vhost.Routes); rLen > 1 && vhost.Routes[rLen-2].Prefix == "/" {
+			catchall := vhost.Routes[rLen-2]
+			vhost.Routes[rLen-2] = vhost.Routes[rLen-1]
+			vhost.Routes[rLen-1] = catchall
+		}
+
+		// TODO: Validate no duplicate path/routes within this virtualhost
+		// (e.g., due to 2 default backends defined in 2 ingresses in different namespaces)
+	}
+
+	// Iterate on all ingresses, and create corresponding rule
+	for _, ingress := range w.ingress.ListIngresses() {
+		for _, rule := range ingress.Spec.Rules {
+			for _, path := range rule.HTTP.Paths {
+				addRule(rule.Host, path.Path, ingress.Namespace, &path.Backend)
+			}
+		}
+		if ingress.Spec.Backend != nil {
+			addRule("", "", ingress.Namespace, ingress.Spec.Backend)
+		}
+	}
+
+	vhosts := vhostMapToSlice(vhostsMap)
+	clusters := clusterMapToSlice(clustersMap)
+
+	listeners := []Listener{}
+	listeners = append(listeners, Listener{
+		Port:           80,
+		BindToPort:     true,
+		UseOriginalDst: true,
```

Comment:
Set to false please. This does not apply to ingress. Its only for internal envoy instances.